pennylane/math/multi_dispatch.py

```diff
@@ -0,0 +1,340 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Multiple dispatch functions"""
+# pylint: disable=import-outside-toplevel,too-many-return-statements
+import warnings
+
+from autoray import numpy as np
+
+from . import single_dispatch  # pylint:disable=unused-import
+from .utils import cast, get_interface
+
+
+def _multi_dispatch(values):
+    """Determines the correct framework to dispatch to given a
+    sequence of tensor-like objects.
+
+    Args:
+        values (Sequence[tensor_like]): a sequence of tensor like objects
+
+    Returns:
+        str: the name of the interface
+
+    To determine the framework to dispatch to, the following rules
+    are applied:
+
+    * Tensors that are incompatible (such as Torch and TensorFlow tensors)
+      cannot both be present.
+
+    * Autograd tensors *may* be present alongside Torch and TensorFlow tensors,
+      but Torch and TensorFlow take precendence; the autograd arrays will
+      be treated as non-differentiable NumPy arrays. A warning will be raised
+      suggesting that vanilla NumPy be used instead.
+
+    * Vanilla NumPy arrays can be used alongside other tensor objects; they will
+      always be treated as non-differentiable constants.
+    """
+    if "resource_variable" in getattr(values, "__module__", tuple()):
+        values = np.asarray(values)
+
+    interfaces = [get_interface(v) for v in values]
+
+    if len(set(interfaces) - {"numpy", "autograd"}) > 1:
+        # contains multiple non-autograd interfaces
+        raise ValueError("Tensors contain mixed types; cannot determine dispatch library")
+
+    non_numpy_interfaces = set(interfaces) - {"numpy"}
+
+    if len(non_numpy_interfaces) > 1:
+        # contains autograd and another interface
+        warnings.warn(
+            f"Contains tensors of types {non_numpy_interfaces}; dispatch will prioritize "
+            "TensorFlow and PyTorch over autograd. Consider replacing Autograd with vanilla NumPy.",
+            UserWarning,
+        )

```

Comment:
Yes, good point. I have to admit, I originally wrote this warning assuming that a user mixing pytorch+tf is very unlikely; users typically pick one ML library and stick to it. However, using pytorch + autograd is much more likely accidentally, since autograd comes with PL.

Reply:
I don't think so no. This is to just warn the user in case they are mixing both autograd arrays and TF arrays. This is very easy to do:
```python
import tensorflow as tf
from pennylane import numpy as np
```

This will work fine, so no reason to raise an error and block it. But switching to vanilla numpy will result in a slight performance boost, so I think a warning is a nice middle ground...